feat: jump to page feature closes #1825

UVLabs opened the pull request:
### Summary
Adds Jump to page pagination feature

### Will affect visual aspect of the product
<!-- It includes visual changes? -->
YES

### Screenshots <!-- if applicable -->
https://vertis.d.pr/3UiAw0

### Test instructions
- Go to Customizer->Layout->Blog / Archive->Scroll down to "Ordering And Content"-> Use checkbox to "Enable Jump To Page" feature or disable.

<!-- Issues that this pull request closes. -->
Closes #1825
<!-- Should look like this: `Closes #1, #2, #3.` . -->

cristian-ungureanu commented on inc/views/pluggable/pagination.php:
```diff
@@ -128,6 +128,115 @@ public function filter_localization( $data ) {
+
+		$current_page = ! empty( get_query_var( 'paged' ) ) ? get_query_var( 'paged' ) : '';
+
+		$markup = <<<MARKUP
```
Please use standard sintax, like `echo '<div id="nv-pagination-jump">';` We don't use heredoc syntax.

UVLabs replied:
@cristian-ungureanu I'm returning the markup here so thats why i didn't use echo, I updated the code to not use heredoc

cristian-ungureanu replied:
@UVLabs, you can collect all your code in a variable and then return it if you need it that way.

cristian-ungureanu commented on inc/views/pluggable/pagination.php:
```diff
@@ -128,6 +128,115 @@ public function filter_localization( $data ) {
+
+		$markup = <<<MARKUP
+		<div id="nv-pagination-jump">
+			<form action="$request" method="GET">
```
All attributes must be escaped. Use `esc_attr` for attributes, `esc_url` for URLs.

UVLabs replied:
For this the `$request` is escaped higher up right after we check if it exists, do we still escape values we put in manually? i.e do we do `method="esc_attr('GET')"` ? and also the same for the div id even when those are hardcoded attributes?

cristian-ungureanu replied:
We don't need to escape hardcoded values like `method="GET"` but we need to escape all the variables when outputting the code. For example if you do `$request = <whatever_escape_function>($request);` and then output it in HTML, the phpcs will throw an error.

cristian-ungureanu commented on inc/views/pluggable/pagination.php:
```diff
@@ -128,6 +128,115 @@ public function filter_localization( $data ) {
+		 * If the next button is present, add the element just before it...
+		 * If it's not then add the jump to element as the last item instead.
+		 */
+		if ( \strpos( $last_element, '<a class="next page-numbers"' ) !== false ) {
```
Use php functions normal, without '\'

cristian-ungureanu reviewed (changes requested):
@UVLabs, have a look at the changes I requested and let me know if there's something unclear.

cristian-ungureanu commented:
The front end doesn't look too good. Probably the field size or some properties from the customizer are applying there and it shouldn't.
<img width="1310" alt="Screenshot 2021-09-11 at 16 21 12" src="https://user-images.githubusercontent.com/9929553/132949319-8d902356-f43a-4791-a7bf-3b767545640b.png">

selul commented:
hey guys, what do you think instead of a new setting we add a new option to this one https://vertis.d.pr/djTfps called Jump to page ?

UVLabs commented:
@cristian-ungureanu Thanks for the review I'll address the changes, the only point that's unclear to me is this: https://github.com/Codeinwp/neve/pull/3083#discussion_r706607249

I checked the source and yes I now see the `paginate_links_output` filter, are you saying instead of doing:

```
if( $this->show_jump_to_page_input() ){
...
}else{
...
}
```

Just do this instead?

```
if( $this->show_jump_to_page_input() ){
add_filter('paginate_links_output', array( $this, 'normalize_jump_to_input' ) );
}
```

Regarding the frontend, I added the padding values based off the default display of neve, I think it may be because of some setting you've changed that affects the size of the page numbers, so it seems like the jump to feature should be linked to that setting as well. 

This is how it currently looks for me after the css is built, I'll have to take a closer look:

![image](https://user-images.githubusercontent.com/10324144/132955616-90e0c72f-cc90-4b65-a12b-7445b8662a1d.png)

---

@selul Yes I think the option would be better placed there indeed.

cristian-ungureanu commented:
@UVLabs you let that code as it was before `$links = paginate_links( array( 'type' => 'list' ) );`, then in the init fuction you add `add_filter('paginate_links_output', array( $this, 'maybe_add_jump_to_input')`  and then you'll have the `maybe_add_jump_to_input($html_markup)` function or however you want to call it where you will check if the input should or should not be added and return $html_markup if not, else parse the HTML and do the changes you want.

cristian-ungureanu commented on inc/views/pluggable/pagination.php:
```diff
@@ -128,6 +128,115 @@ public function filter_localization( $data ) {
+	private function create_jump_to_html() {
+
+		// Value escaped once we confirm it's not empty, we never save this value to the DB.
+		$request = $_SERVER['REQUEST_URI'] ?? ''; //phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
@UVLabs Let's try and get the page number that we've submitted via the $_GET parameter and get the $request URL with this function https://developer.wordpress.org/reference/functions/get_pagenum_link/

UVLabs commented:
@cristian-ungureanu I ran into a dilemma while addressing one of your reviews, you'll let me know how to move forward but it's this one: https://github.com/Codeinwp/neve/pull/3083#discussion_r706608359

If I use the built-in `neve_custom_kses_escape` escaping function we created and I try to echo out the return, then PHPCS throws an output not escaped error, so i'd have to set phpcs to ignore if I am to use it, which from what we've been discussing; shouldn't be done unless necessary, so let me know how I should proceed with that one.

---

Regarding this: https://github.com/Codeinwp/neve/pull/3083#discussion_r708946412 the page number is actually not a GET param in the URL, I removed the attribute all together so it will just use the default, for the Request I switched to using the built-in `$wp->request()` method

cristian-ungureanu reviewed (changes requested):
@UVLabs The code looks great, I've left a few things to update and after that, we can mark this as ready.

@JohnPixle I need your opinion here. We have this pagination https://vertis.d.pr/XHItAb. The problem is that the page numbers are updating their font size from body text, and the Go button from the primary button. We can get to this result https://vertis.d.pr/l6KmSb and that's not that pretty. What do you think we should do?

cc. @selul

UVLabs commented:
> @UVLabs The code looks great, I've left a few things to update and after that, we can mark this as ready.
> 
> @JohnPixle I need your opinion here. We have this pagination https://vertis.d.pr/XHItAb. The problem is that the page numbers are updating their font size from body text, and the Go button from the primary button. We can get to this result https://vertis.d.pr/l6KmSb and that's not that pretty. What do you think we should do?
> 
> cc. @selul

@cristian-ungureanu What if we force it to always respect only the body font size? 

```
#nv-pagination-jump-go {
padding: 0 20px;
font-size: var( --bodyFontSize );
}
```

selul commented:
@abaicus can help too on this matter.

abaicus commented:
@cristian-ungureanu @UVLabs 

The whole form is wrapped inside a <a> tag, which is invalid HTML: https://github.com/Codeinwp/neve/pull/3083/files#diff-57c3e5bb78b1672d2deb8b29ae3447f2eed552652bbb666b199b6b23bce8c143R192-R196

Being a `type=submit` element, wouldn't this inherit all the button properties? Including font size, font weight, line height, colors, etc. The same thing is going to happen to the input, as it uses all the form field variables.

If that is the case, we must make sure we reset all these variables.

I am not a big fan of how this looks. I don't know if it was based on a design, but in my opinion, it should not be intercalated with the current page numbers. 

I'd rather move it to the right side, totally separated from the other links. Here are some examples from dribble:
![image](https://user-images.githubusercontent.com/15010186/135039614-025954da-7425-4781-8507-f67ee8f6b764.png)
![image](https://user-images.githubusercontent.com/15010186/135039874-6c26efd5-5f59-4e22-87bf-ccdc2700308d.png)
![image](https://user-images.githubusercontent.com/15010186/135040005-f1157501-e152-4178-88e8-88c09b819feb.png)

selul commented on inc/views/pluggable/pagination.php:
```diff
@@ -143,6 +144,116 @@ public function filter_localization( $data ) {
+
+		$request = '/' . $wp->request;
+
+		$button_text = apply_filters( 'neve_pagination_jump_button_text', __( 'Go', 'neve' ) );
```
@UVLabs @cristian-ungureanu what do you think about using `greater than` sign instead of text for the button, i.e `&gt;` to avoid adding a new translation.

abaicus commented:
As I didn't answer the initial question:

I think we could simplify this and reduce the impact on the CSS by inlining some JS and using a simple `<a href="#">` tag instead of `input[type=submit]` as this is a pluggable feature that is not enabled by default.

cristian-ungureanu commented:
Thank you for your input @abaicus, I totally agree. There is no design for this, maybe @JohnPixle can help us.

JohnPixle commented:
Hey team, ideally all the typographic elements of the pagination section should be the same size. I had not anticipated that the button styles would apply on the pagination buttons, thanks for pointing this out. Any of the above solutions are fine design-wise, i will leave it to you to evaluate the best technical implementation. Body font size makes sense.

Please keep me in the loop if you need any help on this.

UVLabs commented:
Hey guys @cristian-ungureanu @abaicus I changed the markup and switched to JS instead of using `input[type=submit]` 

Things work better now:

![image](https://user-images.githubusercontent.com/10324144/135185258-b5c34247-48bc-4285-97e4-95fc091eed0a.png)

I pushed to a different branch to keep this one as is:

https://github.com/Codeinwp/neve/blob/feat/jump-to-blog-page-2/inc/views/pluggable/pagination.php
https://github.com/Codeinwp/neve/blob/feat/jump-to-blog-page-2/assets/scss/components/elements/blog/_pagination.scss

@cristian-ungureanu I'll address your reviews once we decide if the new JS approach is preferred over the form approach.

- [ ] TODO: The input field is being stripped on WooCommerce archive page:

![image](https://user-images.githubusercontent.com/10324144/135185720-e493a131-e437-4e65-9c48-4d045f41cf03.png)

abaicus commented:
@UVLabs I'd rather have this all the way to the right, not right next to the other page navigation on desktop. 

Also, let's do what @selul proposed and use `»` or `>` instead of `Jump`.

What I meant by using JS was submitting the form through JS, not doing the whole redirect through it. That redirect will only work with pretty permalinks. 

Furthermore, this throws a notice when using plain permalinks: 
![image](https://user-images.githubusercontent.com/15010186/135223060-ed8ac3da-a62c-4462-a044-b69a01aeba03.png)

I think this should still be a form tag. Right now there is a lot of CSS that is unnecessary and that would be solved by using a proper form tag. We should aim for the smallest SCSS addition we can. The less, the better. A lot of it can be done just by using the aforementioned form tag and rules added only to `#nv-pagination-jump`.

Being a feature that is not enabled by default, I don't think it's necessary to add specific cosmetic rules like these: 
```
input[type="number"]::-webkit-inner-spin-button,
input[type="number"]::-webkit-outer-spin-button {
	opacity: 1;
}
```

> I pushed to a different branch to keep this one as is:

I don't think this is necessary. That's why we have GitHub history 😄

UVLabs commented:
@cristian-ungureanu @abaicus I made the changes and accounted for if site is using plain links

I added support for WooCommerce to the feature as well, and pushed here: https://github.com/Codeinwp/neve-pro-addon/pull/1614

While testing on custom post type archive pages I noticed that the query args were not being rewritten by WordPress, I'm not sure what exactly is happening here but I think it might be because we're using JavaScript to submit the form 

I made a screencast here: https://vertis.d.pr/xwEWON

selul commented on globals/utilities.php:
```diff
@@ -237,6 +239,36 @@ function neve_get_svg_allowed_tags() {
+ *
+ * @return array Extra tags needed by the Jump to Page feature markup that are stripped by wp_kses() by default.
+ */
+function jump_to_page_extra_tags() {
```
This should not be necessary since most of the content that it escapes is hardcoded, not something added by the user or retrieved from the DB. We can escape each of the form dynamic vars in particular and should be fine. 

Also as a sidenote, generally that kind of functions needs to be prefixed.

selul commented on inc/views/pluggable/pagination.php:
```diff
@@ -143,6 +147,155 @@ public function filter_localization( $data ) {
+		$permalink_structure                = get_option( 'permalink_structure' );
+		$plain_permalink_blog_archive_input = ( empty( $permalink_structure ) && $wp_query->is_posts_page ) ? '<input id="nv-page-jump-pid" type="hidden" value="' . esc_attr( $wp_query->query['page_id'] ) . '" name="page_id" />' : '';
+		$plain_permalink_cpt_archive_input  = ( empty( $permalink_structure ) && $wp_query->is_post_type_archive ) ? '<input id="nv-page-jump-cpt" type="hidden" value="' . esc_attr( $wp_query->query['post_type'] ) . '" name="post_type" />' : '';
+		$plain_permalink_cpt_taxonomy_input = ( empty( $permalink_structure ) && $wp_query->is_tax ) ? '<input id="nv-page-jump-cpt-tax" type="hidden" value="' . esc_attr( $wp_query->query_vars['term'] ) . '" name="' . esc_attr( $wp_query->query_vars['taxonomy'] ) . '" />' : '';
```
This is not necessary, You can use https://developer.wordpress.org/reference/functions/get_pagenum_link/ to get the page link

UVLabs replied:
Hey for this can you give me more details about how I'd use the function in this case? The form is submitted to the same page and these input fields are needed for plain permalinks

If im to use that function I can get the number entered in the form from `$_GET` and do a redirect but that would mean the page redirects twice (once when the form is submitted, and second time when we do the redirect) I'm not sure what we should do here

selul commented on inc/views/pluggable/pagination.php:
```diff
@@ -143,6 +147,155 @@ public function filter_localization( $data ) {
+	 */
+	public function expose_jump_to_js() {
+
+		echo "
```
you should use wp_add_inline_script which is recommended for this kind of actions.

UVLabs commented:
@cristian-ungureanu I kept the `a` tag with the inline onclick because  neve automatically tries to style all inputs with type `submit` causing it to look something like this when button styles are changed

<img width="1279" alt="Screenshot 2021-10-04 at 3 42 57 PM" src="https://user-images.githubusercontent.com/10324144/135923123-2bbb6b92-1fd4-44a1-9115-462cbc58d75d.png">

With the way andrei recommended it keeps its form like the other page numbers, I added some styles so it can match what `current` pages use right now:

![image](https://user-images.githubusercontent.com/10324144/135924806-15bf0c52-b7d3-4efb-ba4a-7d54d50a98ca.png)

I think its safe to move this to QA if you agree

cristian-ungureanu commented:
@UVLabs that's not a reason for keeping unnecessary JS. The solution we've found together in our meeting is way better. It doesn't involve JS at all, it's a simple form submit and there's no need to check for permalink type. We can look for a way to exclude that button in CSS.

cristian-ungureanu commented:
Also, please replace the heredoc syntax as we spoke in the meeting.
